compiler/fir/fir2ir/src/org/jetbrains/kotlin/fir/backend/ConversionUtils.kt

```diff
@@ -279,6 +296,64 @@ internal tailrec fun FirCallableSymbol<*>.deepestMatchingOverriddenSymbol(root:
     return overriddenSymbol.deepestMatchingOverriddenSymbol(this)
 }
 
+internal fun IrClass.findMatchingOverriddenSymbolsFromSupertypes(
+    irBuiltIns: IrBuiltIns,
+    target: IrDeclaration,
+    result: MutableList<IrSymbol> = mutableListOf()
+): List<IrSymbol> {
+    for (superType in superTypes) {
+        val superTypeClass = superType.classOrNull
+        if (superTypeClass is IrClassSymbolImpl) {
+            superTypeClass.owner.findMatchingOverriddenSymbolsFromThisAndSupertypes(irBuiltIns, target, result)
+        }
+    }
+    return result
+}
+
+private fun IrClass.findMatchingOverriddenSymbolsFromThisAndSupertypes(
+    irBuiltIns: IrBuiltIns,
+    target: IrDeclaration,
+    result: MutableList<IrSymbol>
+): List<IrSymbol> {
+    for (declaration in declarations) {
+        when {
+            declaration is IrFunction && target is IrFunction ->
+                if (declaration !is IrConstructor &&
+                    !declaration.isFakeOverride &&
+                    declaration.descriptor.modality != Modality.FINAL &&
+                    !Visibilities.isPrivate(declaration.visibility) &&
+                    isOverriding(irBuiltIns, target, declaration)
+                ) {
+                    result.add(declaration.symbol)
+                }
+            // TODO: property lookup to set overriddenSymbols for IrProperty (and accessors)?
+        }
+    }
+    // Stop traversing upwards if we find matching overridden symbols at this level.
+    if (result.isNotEmpty()) {
+        return result
+    }
+    return findMatchingOverriddenSymbolsFromSupertypes(irBuiltIns, target, result)
+}
+
+fun isOverriding(
+    irBuiltIns: IrBuiltIns,
+    target: IrFunction,
+    superCandidate: IrFunction
+): Boolean =
+    target.name == superCandidate.name &&
+            // Not checking the receiver type since we are traversing super type hierarchy, i.e., the expected subtype relation between
+            // receivers is implicitly guaranteed.
+            target.returnType.isSubtypeOf(superCandidate.returnType, irBuiltIns) &&
```

Comment:
This check isn't necessary
```
open class Base {
    fun bar(): String = ""
}

class Derived : Base() {
    // Error: 'bar' hides member of supertype 'Base' ...
    fun bar(): Int = 42
}
```

Reply:
Your example (with proper `open` and `override` modifiers) rather indicates that we need a subtype check for return type. In the current `psi`-based frontend and diagnostic, I got the following error message:
```
Return type is 'Int', which is not a subtype of overridden
public open fun bar(): String defined in Base 
```

Without such check, we're now setting `overriddenSymbol` for `Derived#bar` as `Base#bar`, aren't we? Actually, in your comment, you said: "(they should match each other...". It makes sense that the previous use of subtype didn't handle generic return types well, but I wonder we should at least have `equalTypes` check for return types, which should be extended to cover generic types if needed.

------

With a quick test of using `equalTypes`, though, I saw a number of failing cases (of course), and one of them makes me believe that we need that subtype check for return types due to covariant return types:
```
// In .../box/defaultArguments/function/covariantOverride.kt
open class Foo {
    open fun foo(x: CharSequence = "O"): CharSequence = x
}

class Bar(): Foo() {
    override fun foo(x: CharSequence): String {   // Note the covariant return type
        return x.toString() + "K"
    }
}
```

Reply:
Discussed offline. The example case would be a compilation error anyway, so whether setting `overriddenSymbols` or not doesn't matter.